ENH: Adding where for argmin

m10an opened the pull request:
Addresses [#14371](https://github.com/numpy/numpy/issues/14371)

m10an committed:
Init masking kwargs and testing #14371

m10an committed:
Add `PyArray_BoolArrayConverter` #14371

m10an committed:
Use masking in arg methods #14371

m10an committed:
Update testing and add 2d case #14371

m10an committed:
Add overloads with `initial` and `where` #14371

m10an committed:
Update docs #14371

m10an committed:
Remove `initial` and `where` from `MaskedArray` #14371

m10an commented:
I'm not sure about `np.nanargmin`, `np.nanargmax` and `MaskedArray`'s `argmin`, `argmax`.
Are `initial` and `where` needed in those functions?

m10an committed:
Fix version refrence #14371

m10an committed:
Add new params to `nanfunctions` #14371

m10an committed:
Fix linter warnings #14371

seberg reviewed (commented):
Nice, this is a pretty thorough start!  We need to pass this through the mailing list as an API decision at some point.  I suspect adding `where=` should be pretty straight forward since it matches `max` (and reduce-like operations in general).

OTOH, I don't have any intuition for what `initial=` means.  `initial=` is normally the *starting* value! `np.max([1, 2], initial=100)` makes sense: it returns 100, since that is larger than all the others.
But using `initial` as a "fill" value has a *very* different meaning.  The only way I could make sense of it would be:
```
np.argmax([1, 2], initial=100)
```
returning some special value like `-1` (indicating that `initial` was the largest value.  But that feels like it may be too special, at least unless someone comes around with a clear real-world use case.  (It is much easier/better to do API decisions with a specific use-case, rather than just filling apparent holes in the API)

There are a couple of other issues that I suspect exist that would need fixing here however:
* `where=True` should be the default and be optimized out.
* `where=arr[::2]` i.e. non-contiguous arrays must work and be tested!  (it does not look like they will?)
* `where` must work if shapes mismatch, as long as it can be broadcast to the input.

BvB93 commented on numpy/__init__.pyi:
```diff
@@ -1086,6 +1086,36 @@ class _ArrayOrScalarCommon:
         where: _ArrayLikeBool_co = ...,
     ) -> _NdArraySubClass: ...
 
+    @overload
+    def argmax(
+        self,
+        axis: None = ...,
+        out: None = ...,
+        *,
+        keepdims: bool = ...,
+        initial: _ScalarLike_co = ...,
+        where: _ArrayLikeBool_co = ...,
+    ) -> intp: ...
```
No need for any additional overloads here (or anywhere in this PR), as neither `initial` nor `where` affect the output type, dtype or shape. In this case you can simply add the new parameters to the existing overloads:
``` diff
    @overload
    def argmax(
        self,
        axis: None = ...,
        out: None = ...,
        *,
        keepdims: bool = ...,
+       initial: _ScalarLike_co = ...,
+       where: _ArrayLikeBool_co = ...,
    ) -> intp: ...
```

m10an committed:
ENH: Update `test_masked_2d` (#14371)

m10an committed:
ENH: Add non-contiguous case to `test_masked_2d` (#14371)

m10an committed:
ENH: Fix non-contiguous `where` (#14371)

m10an committed:
ENH: Fix `argmax` and `argmin` overloads (#14371)

As neither `initial` nor `where` affect  the output type,
dtype or shape, they simply should be added to the existing overloads.

m10an commented:
@seberg I totally agree with counter intuitive `initial=` argument. I just followed [example](https://github.com/numpy/numpy/issues/14371#issue-485566842), and realised only while implementing it :)

But `where=` I would use as simple masking, and in case of zero mask raise
```
ValueError: attempt to get argmax of an empty sequence
```
And leave `initial=` as parameter of reduce-functions family

m10an committed:
ENH: Init `where` broadcasting using `NpyIter`

m10an commented:
I've been leisurely working on major issues pointed out by @seberg (non-contiguous case and mismatched shapes).

First one is pretty straight forward (I believe), since there were already a [piece](https://github.com/m10an/numpy/blob/argmin-argmax-masked/numpy/core/src/multiarray/calculation.c#L105) for forcing alignment
```c
wp = (PyArrayObject *)PyArray_ContiguousFromAny((PyObject *)where, NPY_BOOL, 0, 0);
```

But the second one appeared trickier for me. I decided to use `NpyIter` and to not reinvent wheel I thought.
My first test went smoothly ([test_masked](https://github.com/m10an/numpy/blob/argmin-argmax-masked/numpy/core/tests/test_multiarray.py#L4438)), but the second one ([test_masked_2d](https://github.com/m10an/numpy/blob/argmin-argmax-masked/numpy/core/tests/test_multiarray.py#L4448)) exits with `segmentation fault (core dumped)` because of calling of `NpyIter_MultiNew` [here](https://github.com/m10an/numpy/blob/argmin-argmax-masked/numpy/core/src/multiarray/calculation.c#L201).

I've tried to recreate such case using `np.nditer`:
```python
import numpy as np

method = 'max'
n = 5
a = np.zeros([2, n], dtype=int)
where = np.ones([2, n], dtype=bool)
value = getattr(np.iinfo(a.dtype), method)

a[:, 0] = value
a[:, n - 1] = value

arg_method = getattr(a, 'arg' + method)
mask_args = dict(initial=0, where=where)
where[0, 0] = False

it = np.nditer(
    [a, where], 
    ['multi_index'],
    [['readonly'], ['readonly']], casting='no', order='K'
)
with it:
    while not it.finished:
        it.debug_print()
        print(it[0], it[1])
        it.iternext()
```
But that didn't help me much...

The last straw was that gdb suddenly refused to show lines and step through code...
```gdb
Thread 1 "python" hit Breakpoint 1, 0x00007ffff6d0f0f0 in _PyArray_ArgMinMaxCommon () from /home/ivan/proj/m10an/numpy/numpy/core/_multiarray_umath.cpython-39-x86_64-linux-gnu.so
(gdb) list
1       /usr/local/src/conda/python-3.9.13/Programs/python.c: No such file or directory.
(gdb) s
Single stepping until exit from function _PyArray_ArgMinMaxCommon,
which has no line number information.
PyType_IsSubtype (a=0x7ffff70bb200 <PyArray_Type>, b=0x7ffff70d1d80 <PyGenericArrType_Type>) at /usr/local/src/conda/python-3.9.13/Objects/typeobject.c:1425
1425    /usr/local/src/conda/python-3.9.13/Objects/typeobject.c: No such file or directory.
```

I hope there is a stupid mistake that I ignore or maybe `NpyIter` is not best here.
I would appreciate your thought on this.

seberg commented:
Closing this, there is maybe a good start here so can be reopened.  But it needs work and has not been active for 2 years.
